nautobot/extras/tests/test_dynamicgroups.py

```diff
@@ -317,8 +394,259 @@ def test_set_filter(self):
+
+        # Parent should not be here; nested-child should.
+        self.assertNotIn(self.parent, d_flat)
+        self.assertIn(self.nested_child, d_flat)
+
+        # Assert ancestors are deterministic
+        a_tree = self.nested_child.ancestors_tree()
+        a_flat = self.nested_child.flatten_tree(a_tree, descendants=False)
+        expected = {"third-child": 1, "parent": 2}
+        seen = {a.slug: a.depth for a in a_flat}
+        self.assertEqual(seen, expected)
+
+        # Nested-child should not be here; parent should.
+        self.assertNotIn(self.nested_child, a_flat)
+        self.assertIn(self.parent, a_flat)
+
+    def test_ordered_queryset_from_pks(self):
+        """Test `DynamicGroup.ordered_queryset_from_pks()`."""
+        descendants = self.parent.get_descendants()
+        pk_list = [d.pk for d in descendants]
+
+        # Assert that ordering is always deterministic by shuffling the list of pks and asserting
+        # that the ordered pk matches that shuffled order.
+        random.shuffle(pk_list)
+        ordered_qs = self.parent.ordered_queryset_from_pks(pk_list)
+        self.assertEqual(
+            pk_list,
+            [o.pk for o in ordered_qs],
+        )
+
+    def test_generate_query(self):
+        """Test `DynamicGroup.generate_query()`."""
+        # Start with parent. Enumerate descendants and their operators to assert correct results.
+        group = self.parent
+        group_query = group.generate_query()
+        group_qs = group.get_queryset().filter(group_query)
+
+        # <DynamicGroupMembership: First Child: intersection (10)>
+        # <DynamicGroupMembership: Second Child: union (20)>
+        # <DynamicGroupMembership: Third Child: difference (30)>
+        group_members = group.dynamic_group_memberships.all()
+
+        # Manually iterate over the groups to assert the same result that the queryset should have,
+        # igoring weight since each set of members are already ordered by weight when queried from
+        # the database.
+        child_set = set(group.get_queryset())
+        for member in group_members:
+            child_members = set(member.members)
+            operator = member.operator
+            # Use operator value to call a set method; one of intersection, union, difference and
+            # update the results set. (e.g. `child_set.difference(child_members)`
+            if operator == "union":
+                child_set = child_set | child_members
+            elif operator == "difference":
+                child_set = child_set - child_members
+            elif operator == "intersection":
+                child_set = child_set & child_members
+
+        # These should be the same length.
+        self.assertEqual(group_qs.count(), len(child_set))
+
+        # And have the same members...
+        self.assertEqual(
+            sorted(group_qs.values_list("pk", flat=True)),
+            sorted(c.pk for c in child_set),
+        )
```

Comment:
It bugs me that nowhere in these tests do we ever explicitly state what we expect the queryset for `self.parent` to actually evaluate to in terms of specific Devices matched or not matched.... everything is all testing one output calculated by the dynamic group against a different output calculated by the same dynamic group. Good for assuring self-consistency, but I'd feel better if somewhere in the tests we actually had an explicit hard-coded result that we can test against too.